### ISSUE TITLE
Questions about training cost and optimization

### ISSUE THREAD

**Issue by RandyP-P** [arvillion/ActiveVLN#27]
Hi, thanks for your great work.

I have some questions about the training cost, in your paper it says "All experiments are conducted on 4 NVIDIA L40S
GPUs, requiring about 20 hours for IL and RL training respectively.",I used the same base model and training data as you but I am using a single-turn, single-action prediction paradigm, the training cost is byound my expectations(around 90 hours on an 8x A100 (80GB) server).
I am currently fine-tuning the Qwen2.5-VL-3B-Instruct model using the R2R and RxR datasets on an 8x A100 (80GB) server. To evaluate the training efficiency, I ran a 40-step(2560 samples) profiling test and would love to get your thoughts on the estimated training time.
Profiling Details
For my setup, I use a per-device batch size of 8 (global batch size = 64). I used torch.cuda.synchronize() to measure the exact time spent in each forward/backward component. Here is the breakdown:
分段                            平均(ms)       占比     最小(ms)     最大(ms)    采样数
----------------------------------------------------------------------
vit_encode                    2838.0    34.0%     2284.6     3660.7     17
history_compress               263.5     3.2%       12.3     1282.8     17
image_scatter                    3.8     0.0%        3.1        5.7     17
rope_build                       0.5     0.0%        0.3        1.9     17
llm_forward                    629.8     7.5%      577.1      806.2     17
backward_only                 4410.0    52.8%
data+framework+other           213.0     2.5%
----------------------------------------------------------------------
平均 step 总耗时                   8358.6 ms

图片数(输入vit的图片数):     min=99, max=158, avg=121.9
序列长度(进入LLM的token数):   min=2097, max=2535, avg=2346.3

With synchronization turned off, a single step takes about 8 seconds.there's about 2.4M samples in r2r and rxr datasets, in terms of this calculation method it takes about 90 hours (or ~720 GPU hours), which feels quite long.
My Questions
1.Did you use the complete R2R and RxR datasets for the IL training phase?
2.Did you apply any specific operations to optimize the training time?
3.Based on my profiling breakdown and the training config below, do you have any recommendations or best practices to optimize the training speed?
Thank you so much for your help!

My training configuration is as follows:
torchrun --nproc_per_node=$NPROC_PER_NODE \
            --master_addr=$MASTER_ADDR \
            --master_port=$MASTER_PORT \
            qwen-vl-finetune/qwenvl/train/train_qwen.py \
            --model_name_or_path $MODEL_PATH \
            --tune_mm_llm True \
            --tune_mm_vision False \
            --tune_mm_mlp True \
            --dataset_use $DATASETS \
            --output_dir $OUTPUT_DIR \
            --cache_dir $CACHE_DIR \
            --bf16 \
            --per_device_train_batch_size 8 \
            --gradient_accumulation_steps 1 \
            --learning_rate 2e-5 \
            --optim adamw_torch \
            --model_max_length 8192 \
            --data_flatten False \
            --max_pixels $((576*28*28)) \
            --min_pixels $((16*28*28)) \
            --base_interval 2 \
            --video_max_frames 8 \
            --video_min_frames 4 \
            --num_train_epochs 1 \
            --warmup_ratio 0.03 \
            --lr_scheduler_type "cosine" \
            --weight_decay 0.01 \
            --logging_steps 10 \
            --save_steps 1000 \
            --save_total_limit 1 \
            --deepspeed "qwen-vl-finetune/scripts/zero2.json" \
            --gradient_checkpointing \
            --dataloader_num_workers 8 \
            --seed 42 \
            --logging_dir "${OUTPUT_DIR}/runs" \
            --report_to "tensorboard" \
            --max_steps 40 \
            2>&1 | tee "${OUTPUT_DIR}/train.log"

**Comment by waynechu1021** [comment]
> Hi, thanks for your great work.
> 
> ## I have some questions about the training cost, in your paper it says "All experiments are conducted on 4 NVIDIA L40S
> GPUs, requiring about 20 hours for IL and RL training respectively.",I used the same base model and training data as you but I am using a single-turn, single-action prediction paradigm, the training cost is byound my expectations(around 90 hours on an 8x A100 (80GB) server).
> I am currently fine-tuning the Qwen2.5-VL-3B-Instruct model using the R2R and RxR datasets on an 8x A100 (80GB) server. To evaluate the training efficiency, I ran a 40-step(2560 samples) profiling test and would love to get your thoughts on the estimated training time.
> Profiling Details
> For my setup, I use a per-device batch size of 8 (global batch size = 64). I used torch.cuda.synchronize() to measure the exact time spent in each forward/backward component. Here is the breakdown:
> 分段 平均(ms) 占比 最小(ms) 最大(ms) 采样数
> ## vit_encode 2838.0 34.0% 2284.6 3660.7 17
> history_compress 263.5 3.2% 12.3 1282.8 17
> image_scatter 3.8 0.0% 3.1 5.7 17
> rope_build 0.5 0.0% 0.3 1.9 17
> llm_forward 629.8 7.5% 577.1 806.2 17
> backward_only 4410.0 52.8%
> data+framework+other 213.0 2.5%
> 平均 step 总耗时 8358.6 ms
> 
> 图片数(输入vit的图片数): min=99, max=158, avg=121.9 序列长度(进入LLM的token数): min=2097, max=2535, avg=2346.3
> 
> With synchronization turned off, a single step takes about 8 seconds.there's about 2.4M samples in r2r and rxr datasets, in terms of this calculation method it takes about 90 hours (or ~720 GPU hours), which feels quite long. My Questions 1.Did you use the complete R2R and RxR datasets for the IL training phase? 2.Did you apply any specific operations to optimize the training time? 3.Based on my profiling breakdown and the training config below, do you have any recommendations or best practices to optimize the training speed? Thank you so much for your help!
> 
> My training configuration is as follows: torchrun --nproc_per_node=$NPROC_PER_NODE --master_addr=$MASTER_ADDR --master_port=$MASTER_PORT qwen-vl-finetune/qwenvl/train/train_qwen.py --model_name_or_path $MODEL_PATH --tune_mm_llm True --tune_mm_vision False --tune_mm_mlp True --dataset_use $DATASETS --output_dir $OUTPUT_DIR --cache_dir $CACHE_DIR --bf16 --per_device_train_batch_size 8 --gradient_accumulation_steps 1 --learning_rate 2e-5 --optim adamw_torch --model_max_length 8192 --data_flatten False --max_pixels $((576_28_28)) --min_pixels $((16_28_28)) --base_interval 2 --video_max_frames 8 --video_min_frames 4 --num_train_epochs 1 --warmup_ratio 0.03 --lr_scheduler_type "cosine" --weight_decay 0.01 --logging_steps 10 --save_steps 1000 --save_total_limit 1 --deepspeed "qwen-vl-finetune/scripts/zero2.json" --gradient_checkpointing --dataloader_num_workers 8 --seed 42 --logging_dir "${OUTPUT_DIR}/runs" --report_to "tensorboard" --max_steps 40 2>&1 | tee "${OUTPUT_DIR}/train.log"

With the same number of trajectories, a single-turn dialogue setup, predicting only one action at a time, will generate several times more training samples. Besides, I don't quite understand your setting of `图片数(输入vit的图片数): min=99, max=158, avg=121.9`.

**Comment by RandyP-P** [comment]
Thank you for your reply, I have figured out the reason for the training speed discrepancy. It is mainly because you used action fusion and multi-step prediction, which reduced the number of supervised training samples
